### PR TITLE
Simple observable collection optimizations

### LINKED ISSUE #134: Use of ObservableCollection<T>.Remove(T)
As first stated in https://github.com/elmish/Elmish.WPF/issues/131#issuecomment-536340393, I think the use of [`ObservableCollection<T>.Remove(T)`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.objectmodel.collection-1.remove?view=netframework-4.8) (in three calls in `updateValue`) is a bug.  Instead of removing by item, I think the correct implementation is to remove by index via [`ObservableCollection<T>.RemoveAt(Int32)`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.objectmodel.collection-1.removeat?view=netframework-4.8).

This issue will serve as the tracking of this (potential) problem (instead of #131...that issue has enough going on in it :P ).

Here is the relevant part of that previous comment.

------------

Recall the update logic for the `OneWaySeq` case of `updateValue`.
https://github.com/elmish/Elmish.WPF/blob/8f78f833bef5c903e6947b23688cef30480ccf8b/src/Elmish.WPF/ViewModel.fs#L341-L343

I think that call to `Remove` contains a bug.  The precise thing to do is remove by index.  Removing by item means that the collection needs to find which item to remove.  The documentation for `ObservableCollection<T>.Remove` [says](https://docs.microsoft.com/en-us/dotnet/api/system.collections.objectmodel.collection-1.remove?view=netframework-4.8#remarks) it removes the first item it finds for which that item and the given item return `true` when passed into the default equality comparer of type `T`.  If `itemEquals` is not the same as this default equality comparer, then this call to `Remove` will not have the intended behavior.  I can create a test that exhibits this problem and a PR with a fix.

### PR DESCRIPTION
**Pull request by bender2k14**
Fixes #134 and partially addresses #137 by implementing all of the "simple" optimizations.

I have not tested anything yet.  I will test later.

The two remaining more complicated optimizations are removing all elements that need to be removed in nearly linear time (as [discussed here](https://github.com/elmish/Elmish.WPF/issues/137#issue-500916222)) and ordering in nearly linear time (as [discussed here](https://github.com/elmish/Elmish.WPF/issues/137#issuecomment-537744117)).

### CODE REVIEW

**Comment by cmeeren** [comment]
Thanks a lot! Let me know when you're ready to merge.

@reinux do you want to give this PR a spin to check how much it impacts the performance you mentioned in https://github.com/elmish/Elmish.WPF/issues/113#issuecomment-541227271?

**Comment by reinux** [comment]
Works great!
Down from about 250ms per update to 120ms.
Haven't set up any proper tests, but it seems to work as expected.

**Comment by cmeeren** [comment]
@reinux @bender2k14 Do you think the dictionary stuff in https://github.com/elmish/Elmish.WPF/issues/137#issuecomment-541247319 could work well in this PR?

**Comment by reinux** [comment]
The `Seq.find` and the `Move` call that follows are actually what causes the bug that I mentioned in that comment. As @bender2k14 notes, caching the ID->ixs of the current items up front causes it to be out of sync as the `Move` calls change the indices. His branch actually uses the dictionary for the rest of the function, where that isn't an issue.

**Comment by bender2k14** [comment]
Indeed.  Of all the planning that I did for this PR, I didn't notice that replacing `Seq.fine` with dictionary lookups would cause a bug.  Good thing @reinux ran into it before me ;) :P

**Comment by bender2k14** [comment]
I added a commit that calls `Clear` when it would be efficient to do so

**Comment by cmeeren** [comment]
Seems that commit broke some tests. Could you investigate?

**Comment by bender2k14** [comment]
Oh, I had an improvement to that commit that I forgot to include.  I force pushed a new commit.  The tests all pass for me.

The previous commit had the correct behavior, but raised more (clear) events than it needed to.  It is great to see that there are some tests that noticed the difference.

**Comment by bender2k14** [comment]
> Thanks a lot

My pleasure :)

> Let me know when you're ready to merge.

I am ready.  I have reviewed my changes several times and have done some manual testing.  I think everything is working.

> PS, you can mark PRs as "drafts" or something like that when you create them, to signal if you don't want them merged right away.

Oh, great.  I will keep that in mind for next time.

**Comment by cmeeren** [comment]
In https://github.com/elmish/Elmish.WPF/issues/134#issuecomment-541492334 you mention adding a test; should that be part of this PR or should I go ahead and merge?

**Comment by bender2k14** [comment]
It makes sense to add the test to this PR.  Coming soon!

**Comment by bender2k14** [comment]
It took me longer than I expected to write the test.  The good news though is that I think it is complete.  I have it failing on `master`.  Tomorrow I will confirm that it passes on the breach in this PR.  Then I will force push the branch with the test created first.

**Comment by bender2k14** [comment]
Done! :D

Ready to merge

**Comment by cmeeren** [comment]
Thanks a lot! Will try to review and merge tomorrow.

File: src/Elmish.WPF.Tests/ViewModelTests.fs [hunk 1]
```diff
@@ -818,6 +815,87 @@ module OneWaySeqLazy =
...
+  type TestClass (id: int, data: string) =
+    member _.Id = id
+    member _.Data = data
+    override __.GetHashCode() = 0
+    override __.Equals that =
+      // All instances of TestClass are considered equal.
+      // Not very helpful, but a valid implementation.
+      match that with
+      | null -> false
+      | :? TestClass as x -> true
```
Comment (cmeeren) [inline src/Elmish.WPF.Tests/ViewModelTests.fs]: Actually, you can remove the `null` check. Type checks à la `:? TestClass` will only succeed if it's not null.

Comment (bender2k14) [reply src/Elmish.WPF.Tests/ViewModelTests.fs]: Sure.  Force pushed that change.

**Pull request merged**